Sprint 07

AlbertKhaydarov opened the pull request:
Написаны и проходят unit-тесты:
На сабскрипт для безопасного получения элемента массива по индексу (ArrayTests);
На сервис MoviesLoader (для проверки успешной загрузки и для проверки ошибки MoviesLoaderTests).
Написаны и проходят UI-тесты:
Два теста на замену одного постера на другой при нажатии кнопки ответа — 
Тест кнопки Да (при нажатии на кнопку Да меняется лейбл с номером вопроса).
Тест кнопки Нет (при нажатии на кнопку Нет меняется лейбл с номером вопроса).
Тест появления алерта при окончании раунда ( проверка, что алерт появился, тест текста на кнопке и в заголовке алерта).
Тест скрытия алерта после нажатия на кнопку на нём (в этом тесте  проверяется, что алерт отсутствует на экране и счётчик вопросов сбросился к значению 1/10).
Проведён рефакторинг проекта — отрефакторен MovieQuizViewController на архитектуру MVP по шагам из 3 и 4 уроков темы 5 с учетом своей реализации (отличается от показанной в уроке).
Написан и проходит тест для convert(model: QuizQuestion) -> QuizStepViewModel.

AlbertKhaydarov committed:
уроки тестирования

AlbertKhaydarov committed:
Урок по тестированию

AlbertKhaydarov committed:
Сделал тесты из урока

AlbertKhaydarov committed:
вынес метод convert() из MoveQuizViewController в Presenter

AlbertKhaydarov committed:
Перенес логику yesButtonClicked(_ sender: UIButton) , noButtonClicked(_ sender: UIButton)  в presenter

AlbertKhaydarov committed:
Перенес showNextQuestionOrResults в Presenter

AlbertKhaydarov committed:
Перенес в presenter реализацию QuestionFactoryDelegate, ResultAlertPresenterDelegate, ErrorAlertPresenterDelegate

AlbertKhaydarov committed:
Перенес в presenter showAnswerResult(isCorrect: Bool

AlbertKhaydarov committed:
Сделан MovieQuizViewControllerProtocol (подписан  MovieQuizViewController, изменена зависимость viewController на протокол в presenter), сделан testPresenterConvertModel

So1ovyov commented on MovieQuizUITests/MovieQuizUITestsLaunchTests.swift:
```diff
@@ -0,0 +1,32 @@
+
+import XCTest
+
+//final class MovieQuizUITestsLaunchTests: XCTestCase {
```
В целом этот файл можно удалить.

AlbertKhaydarov replied:
Выполнено

So1ovyov commented on MovieQuizTests/MoviesLoaderTests.swift:
```diff
@@ -0,0 +1,108 @@
+import XCTest // не забывайте импортировать фреймворк для тестирования
+@testable import MovieQuiz // импортируем приложение для тестирования
+
+struct StubNetworkClient: NetworkRouting {
```
Можно вынести это в отдельный файл вспомогательный для теста.

AlbertKhaydarov replied:
Выполнено

So1ovyov commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,72 +1,52 @@
+        presenter.noButtonClicked()
     }
     
     // MARK: - Private functions
```
Написано Private но дальше идут методы публичные

AlbertKhaydarov replied:
Выполнено

So1ovyov commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,161 @@
+}
+
+// MARK: - QuestionFactoryDelegate
+extension MovieQuizPresenter: QuestionFactoryDelegate {
```
Круто что реализации протоколов выделил в отдельные расширения. Молодец !

AlbertKhaydarov committed:
Удален файл MovieQuizUITests/MovieQuizUITestsLaunchTests.swift
Сделан отдельный файл struct StubNetworkClient с данными для тестирования
Изменен уровень доступности func yesButtonClicked и func noButtonClicked на internal, т.к. используют внутри не приватные методы
